### PR TITLE
Skip self address in dialing

### LINKED ISSUE #773: Automatically refuse connections coming from the local node
Right now I don't think this is enforced anywhere.

A test should also be written.

### PR DESCRIPTION
**Pull request by l4l**
Resolves #773 

It have O(n^2) complexity, though as number of peer addresses is usually considered small, that shouldn't be a problem, should it? Also it might be handy to use [itertools](https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.cartesian_product) crate, but that might be too much for just one function. Also I've spent for about 10 minutes lurking the code so I'm completely unfamiliar with it (and might miss other important changes).

p.s: hey from rustrush

### CODE REVIEW

File: core/src/swarm.rs [hunk 1]
```diff
@@ -164,12 +164,18 @@ where TBehaviour: NetworkBehaviour<TTopology>,
...
 
     /// Tries to reach the given peer using the elements in the topology.
+    /// Silently returns if address was associated with the `Self`
     ///
     /// Has no effect if we are already connected to that peer, or if no address is known for the
     /// peer.
     #[inline]
     pub fn dial(me: &mut Self, peer_id: PeerId) {
         let addrs = me.topology.addresses_of_peer(&peer_id);
+        for a in addrs.iter() {
```
Comment (dvdplm) [inline core/src/swarm.rs]: Maybe use `any`? (untested!)
```suggestion
       let is_dialling_self = addrs.iter().any(|addr| {
            me.listened_addrs.iter().any(|my_addr| my_addr == addr)
        })
        if is_dialling_self {
            trace!("can't dial self!");
            return;
         }
```

Comment (l4l) [reply core/src/swarm.rs]: Yeah, didn't notice that function, thanks!

**Comment by tomaka** [comment]
> p.s: hey from rustrush

Hey!

Thanks for the PR!
However I don't think the change should be in `swarm.rs`. I think that it should rather be in `raw_swarm.rs`, as this is the entry point for all dialing.
I realize that the code in `raw_swarm.rs` is much more complex though, and not as easy to fix.

**Comment by l4l** [comment]
> However I don't think the change should be in swarm.rs. I think that it should rather be in raw_swarm.rs, as this is the entry point for all dialing.

Indeed, I've noticed `dial_addr` above that uses `raw_swarm`. Also it seems that `RawSwarm` is only used at _swarm.rs_, so it is an internal structure, right? So it might be possible to handle that behavior at `Swarm::dial` & `Swarm::dial_addr`. Is there any problem with such a way?

If no, I still would like to collaborate on the issue and find the possible solution

**Comment by tomaka** [comment]
It is the responsibility of the `raw_swarm` to handle all the connections. The `swarm` is just a convenient wrapper on top of that. It is therefore not the responsibility of the swarm to handle that.

**Comment by tomaka** [comment]
Sorry, the fix was starting to be pressing so we did #839 instead

**Pull request closed without merging**